### PR TITLE
feat(cmd/influx): allow setting shard-group durations for buckets via API and CLI

### LINKED ISSUE #19764: Add shard duration settings to the CLI
based on the outcome of this https://github.com/influxdata/influxdb/issues/19738

Let's add the ability to adjust bucket shards from the CLI.

### PR DESCRIPTION
**Pull request by danxmoran**
Closes #19764 

Builds off of #20620 to address review feedback and ensure shard-group duration is tracked in metadata, so the CLI can report on it in all cases.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ the endpoint has been removed. Use the `/metrics` endpoint to collect system sta
 1. [20827](https://github.com/influxdata/influxdb/pull/20827): Add `--pprof-disabled` option to `influxd` to disable exposing profiling information over HTTP.
 1. [20827](https://github.com/influxdata/influxdb/pull/20827): Add `/debug/pprof/all` HTTP endpoint to gather all profiles at once.
 1. [20827](https://github.com/influxdata/influxdb/pull/20827): Upgrade `http.pprof-enabled` config in `influxd upgrade`.
+1. [20911](https://github.com/influxdata/influxdb/pull/20911): Add support for explicitly setting shard-group durations on buckets. Thanks @hinst!
```
Comment (danxmoran) [inline CHANGELOG.md:21]: Would it be more clear to call it "shard duration" everywhere? Some of the errors from the copied-over v1 code use that wording.

Comment (lesam) [reply CHANGELOG.md:21]: So as I understand it:

`shard duration` == `shard group duration` is the amount of time each shard covers.

`duration` (of a retention policy) is the amount of time we actually keep the data for.

If we were starting over I'd want to call the first thing `shard span` or some other word, and the second thing the `duration`.

If we can't do that maybe we can standardize on 'shard group duration' and 'retention duration'?

@sanderson might have a recommendation

File: tenant/http_client_bucket.go [hunk 2]
```diff
@@ -168,7 +161,7 @@ func (s *BucketClientService) UpdateBucket(ctx context.Context, id influxdb.ID,
 	if err != nil {
 		return nil, err
 	}
-	return br.toInfluxDB()
+	return br.toInfluxDB(), nil
```
Comment (lesam) [inline tenant/http_client_bucket.go:164]: Is the PatchJSON call being called in the tests above?

File: tenant/http_server_bucket.go [hunk 3]
```diff
@@ -149,25 +132,42 @@ func newBucket(pb *influxdb.Bucket) *bucket {
...
+			Msg:  "buckets cannot have more than one retention rule at this time",
+		}
+	}
+
 	if len(b.RetentionRules) > 0 {
-		_, err := b.RetentionRules[0].RetentionPeriod()
-		if err != nil {
-			return err
+		rule := b.RetentionRules[0]
+		if rule.EverySeconds != nil && *rule.EverySeconds < 0 {
```
Comment (lesam) [inline tenant/http_server_bucket.go:163]: Should we check Shard Group Duration is non-negative as well?

File: tenant/http_client_bucket.go (hunk 2 above)
Comment (danxmoran) [reply tenant/http_client_bucket.go:164]: Yes, in `cmd/influxd/launcher/storage_test.go`. The `BucketService` created [here](https://github.com/influxdata/influxdb/pull/20911/files#diff-77b53aadc718021b69a618cf71d07460e439078ec62376689176beed136868f0R275) is a [wrapper](https://github.com/influxdata/influxdb/blob/master/cmd/influxd/launcher/launcher_helpers.go#L383) around the HTTP client.

File: tenant/http_server_bucket.go (hunk 3 above)
Comment (danxmoran) [reply tenant/http_server_bucket.go:163]: Done, and added tests for it.

File: tenant/http_server_bucket_test.go [hunk 4]
```diff
@@ -78,3 +82,133 @@ func initBucketHttpService(f itesting.BucketFields, t *testing.T) (influxdb.Buck
...
+			wants: wants{
+				err: &influxdb.Error{
+					Code: influxdb.EUnprocessableEntity,
+					Msg:  "shard-group duration seconds cannot be negative",
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (danxmoran) [inline tenant/http_server_bucket_test.go:185]: Lazily copy-pasted from `testing/bucket_service.go`

File: cmd/influx/bucket.go [hunk 5]
```diff
@@ -320,7 +339,7 @@ func (b *cmdBucketBuilder) printBuckets(printOpt bucketPrintOpt) error {
 
 	w.HideHeaders(b.hideHeaders)
 
-	headers := []string{"ID", "Name", "Retention", "Organization ID"}
+	headers := []string{"ID", "Name", "Retention", "Shard duration", "Organization ID"}
```
Comment (lesam) [inline cmd/influx/bucket.go:342]: Shard group duration?

File: cmd/influx/bucket.go [hunk 6]
```diff
@@ -331,10 +350,22 @@ func (b *cmdBucketBuilder) printBuckets(printOpt bucketPrintOpt) error {
...
+		if bkt.ShardGroupDuration == 0 {
+			sgDur = "n/a"
+		}
+
 		m := map[string]interface{}{
 			"ID":              bkt.ID.String(),
 			"Name":            bkt.Name,
-			"Retention":       bkt.RetentionPeriod,
+			"Retention":       rp,
+			"Shard duration":  sgDur,
```
Comment (lesam) [inline cmd/influx/bucket.go:368]: same?

File: cmd/influx/bucket.go (hunk 5 above)
Comment (danxmoran) [reply cmd/influx/bucket.go]: Fixed

File: cmd/influx/bucket.go (hunk 6 above)
Comment (danxmoran) [reply cmd/influx/bucket.go]: Fixed

**Review by lesam — APPROVED** [review]
LGTM. This feels like more change than I would have expected for piping through a single API field (especially all the stuff in http_server_bucket), which worries me a little.

**Pull request merged**